src/qt/pivx/coldstakingmodel.cpp

```diff
@@ -59,9 +58,26 @@ void ColdStakingModel::refresh() {
                 del.delegatedUtxo.unite(delegation.delegatedUtxo);
                 del.cachedTotalAmount += delegation.cachedTotalAmount;
             }
-            cachedAmount += delegation.cachedTotalAmount;
+
         }
     }
+    cachedAmount = getUpdatedTotalAmount();
+}
+
+/**
+ * updates cachedAmount from cachedDelegations and returns it
+ */
+CAmount ColdStakingModel::getUpdatedTotalAmount()
+{
+    cachedAmount = 0;
+    for (const CSDelegation& delegation : cachedDelegations) {
+        // add amount to cachedAmount if either:
+        // - this is a owned delegation
+        // - this is a staked delegation, and the owner is whitelisted
+        if (!delegation.isSpendable && !addressTableModel->isWhitelisted(delegation.ownerAddress)) continue;
+        cachedAmount += delegation.cachedTotalAmount;
+    }
+    return cachedAmount;
```

Comment:
This will be a problem with lot of p2cs uxto. It's running in the main UI thread.

You can update the cachedAmount in every whitelist/blacklist without need to loop over the cached delegations.

whitelist ---> cachedAmount += whitelistedAmount.
blacklist ---> cachedAmount -= blacklistedAmount.

But well, if you do that, you will have to update the cached amount when the user uses the rpc calls to whitelist/blacklist delegations too.

Reply:
Minor issue. The whitelist/blacklist is a "una tantum" operation, not something to do often. And if the staker has many p2cs, it's quite likely that the owner has been already whitelisted.
Nonetheless, better to have it more efficient.